clippy_lints/src/returns.rs

```diff
@@ -353,23 +369,73 @@ fn check_final_expr<'tcx>(
+            check_block_return(
+                cx,
+                &then.kind,
+                peeled_drop_expr.span,
+                semi_spans.clone(),
+                parent_fn_return_type,
+            );
             if let Some(else_clause) = else_clause_opt {
-                check_block_return(cx, &else_clause.kind, peeled_drop_expr.span, semi_spans);
+                check_block_return(
+                    cx,
+                    &else_clause.kind,
+                    peeled_drop_expr.span,
+                    semi_spans,
+                    parent_fn_return_type,
+                );
             }
         },
         // a match expr, check all arms
         // an if/if let expr, check both exprs
         // note, if without else is going to be a type checking error anyways
         // (except for unit type functions) so we don't match it
-        ExprKind::Match(_, arms, MatchSource::Normal) => {
-            let match_ty = cx.typeck_results().expr_ty(peeled_drop_expr);
-            for arm in *arms {
-                check_final_expr(cx, arm.body, semi_spans.clone(), RetReplacement::Unit, Some(match_ty));
+        ExprKind::Match(match_expr, arms, MatchSource::Normal) => {
+            // need some special checks for cases where we would try to convert
+            // match statement to match expression
+            let check_arms = if parent_stmt_kind_opt.is_some_and(|x| matches!(x, StmtKind::Semi(_))) {
+                let match_expr_ret_type = cx.typeck_results().expr_ty(match_expr);
+
+                // check if function return type is equal to match return type
+                // if yes then we want can convert this match stmt to match expr
+                // otherwise dont do it
+                match parent_fn_return_type {
+                    Some(FnRetTy::DefaultReturn(_)) if let ty::Tuple(tuple) = match_expr_ret_type.kind() =>
+                        tuple.len() == 0,
+                    Some(FnRetTy::Return(_ret_ty)) => matches!(match_expr_ret_type.kind(), _ret_ty),
+                    _ => false,
+                }
+            } else {
+                // we always want to check arms in case match is already an expression
+                true
+            };
+
+            if check_arms {
+                arms.iter().for_each(|arm| {
+                    check_final_expr(
+                        cx,
+                        arm.body,
+                        semi_spans.clone(),
+                        RetReplacement::Unit,
+                        // Here we pass this as arm could have
+                        // - match directly as expr: in this case our thing works
+                        // - block expr: in this case our thing gets overriden
+                        // by call to check_block_return
+                        // - anything else: doesn't matter
+                        Some(StmtKind::Expr(arm.body)),
```

Comment:
Is this any different than passing `None`?

Reply:
Heyo!

Yup nice catch it shouldn't matter, pushing a change for the same :)